script_node_communication/main/app_main.cpp

```diff
@@ -15,54 +14,38 @@ extern "C"
 
 void run_demo()
 {
-    uint8_t req_success = 0;
-
     dai::SpiApi mySpiApi;
     mySpiApi.set_send_spi_impl(&esp32_send_spi);
     mySpiApi.set_recv_spi_impl(&esp32_recv_spi);
 
     sleep(5);
 
-    while (1)
-    {
-        dai::Message nnMsg;
-        if(mySpiApi.req_message(&nnMsg, "spimetaout")){
-            dai::RawNNData nnData;
-            mySpiApi.parse_metadata(&nnMsg.raw_meta, nnData);
-            printf("Tensors size: %u\n",nnData.tensors.size());
-            for(int i=0; i<nnData.tensors.size(); i++) {
-            // for(const auto& tensor : nnData.tensors){
-                auto tensor = nnData.tensors[i];
-                printf("============= i=%u ===========", i);
-                printf("Dimensions %u\n", tensor.numDimensions);
-                size_t size = tensor.dims[0] * tensor.strides[0];
-                printf("name of the sensor :%s\n",tensor.name.c_str());
-                printf("size of the tensor: %u\n", size);
-                auto beg = nnData.data.begin() + tensor.offset;
-                auto end = beg + size;
-                size_t data_size = nnData.data.size();
-                printf("size of the data :%d, ptr: %08X\n", nnMsg.raw_data.size, (uint32_t) nnMsg.raw_data.data);
+    // Create RawBuffer message and set some text to it
+    dai::RawBuffer buf;
+    std::string str = "Hello";
+    std::vector<uint8_t> strVec(str.begin(), str.end());
+    buf.data = strVec;
+
+    // Send message and wait for the return message
+    dai::Message msg;
+    while(!mySpiApi.req_message(&msg, "ret_str")) {
+        std::cout << "Sending string '" << str << "' to VPU..." << std::endl;
+        mySpiApi.send_message(buf, "str");
+        sleep(1);
+    }
 
-                printf("msg: %d\n", nnMsg.raw_data.data[0]);
+    std::string retStr(msg.raw_data.data, msg.raw_data.data + msg.raw_data.size);
+    std::cout << "String recieved from VPU: '" << retStr << "'" << std::endl;
 
-            }
-            // free up resources once you're done with the message.
-            mySpiApi.free_message(&nnMsg);
-            mySpiApi.spi_pop_messages();
-        } else {
-            // Periodically check if new messages are available
-            usleep(100);
-        }
```

Comment:
Don't drop this - retain so polling is limited

Reply:
what do you mean? On line 34 there's `sleep(1);`

Reply:
If requesting message fails, then it'll loop right back and poll device again immediately. To reduce this polling, consider sleeping for a short amount of time in between

Reply:
Nevermind missed the not operator there. LGTM